packages/gatsby/src/commands/build-html.ts

```diff
@@ -175,25 +179,71 @@ const renderHTMLQueue = async (
       ? workerPool.renderHTMLProd
       : workerPool.renderHTMLDev
 
-  await Bluebird.map(segments, async pageSegment => {
-    await renderHTML({
-      envVars,
-      htmlComponentRendererPath,
-      paths: pageSegment,
-      sessionId,
-    })
+  const uniqueUnsafeBuiltinUsedStacks = new Set<string>()
+
+  try {
+    await Bluebird.map(segments, async pageSegment => {
+      const htmlRenderMeta: IRenderHtmlResult = await renderHTML({
+        envVars,
+        htmlComponentRendererPath,
+        paths: pageSegment,
+        sessionId,
+      })
 
-    if (stage === `build-html`) {
+      if (stage === `build-html`) {
+        store.dispatch({
+          type: `HTML_GENERATED`,
+          payload: pageSegment,
+        })
+
+        for (const [_pagePath, arrayOfUsages] of Object.entries(
+          htmlRenderMeta.unsafeBuiltinsUsageByPagePath
+        )) {
+          for (const unsafeUsageStack of arrayOfUsages) {
+            uniqueUnsafeBuiltinUsedStacks.add(unsafeUsageStack)
+          }
+        }
+      }
+
+      if (activity && activity.tick) {
+        activity.tick(pageSegment.length)
+      }
+    })
+  } catch (e) {
+    if (e?.context?.unsafeBuiltinsUsageByPagePath) {
+      for (const [_pagePath, arrayOfUsages] of Object.entries(
+        e.context.unsafeBuiltinsUsageByPagePath
+      )) {
+        // @ts-ignore TS doesn't know arrayOfUsages is Iterable
+        for (const unsafeUsageStack of arrayOfUsages) {
+          uniqueUnsafeBuiltinUsedStacks.add(unsafeUsageStack)
+        }
+      }
+    }
+    throw e
+  } finally {
+    if (uniqueUnsafeBuiltinUsedStacks.size > 0) {
+      console.warn(
+        `Unsafe builtin method was used, future builds will need to rebuild all pages`
+      )
       store.dispatch({
-        type: `HTML_GENERATED`,
-        payload: pageSegment,
+        type: `SSR_USED_UNSAFE_BUILTIN`,
```

Comment:
How does someone clear this flag? Running `gatsby clean` and/or gatsby cleans cache?

Reply:
this flag gets cleared when SSR compilation hash changes (as long as we use same SSR compilation hash is used - we know that if previous build used unsafe builtin methods - following builds will continue to do so, because it's the same code running).

Changing SSR compilation hash also cause all pages to be rebuild so we are "covered" for that case